Kernel/Term.cpp

```diff
@@ -1920,33 +1912,59 @@ TermList Term::typeArg(unsigned n) const
 }
 
 bool Term::computable() const {
-  if (!env.signature->getFunction(this->functor())->computable()) {
-    return false;
-  }
-  SubtermIterator sit(this);
-  while (sit.hasNext()) {
-    TermList t = sit.next();
-    if (!t.isTerm() || !env.signature->getFunction(t.term()->functor())->computable()) {
+  return ground() && computableOrVar();
+}
+
+bool Term::computableOrVarHelper(DHMap<unsigned, unsigned>* recAncestors) const {
+  Signature::Symbol* symbol = env.signature->getFunction(functor());
+  Shell::SynthesisALManager* synthMan = static_cast<Shell::SynthesisALManager*>(Shell::SynthesisALManager::getInstance());
+
+  if (!synthMan->isFunctionComputable(functor())) {
+    // either an uncomputable symbol from the input, or an introduced symbol
+    if (!symbol->skolem()) { // computability of skolems depends on the context, all else is really uncomputable
       return false;
     }
+    if (!synthMan->isRecTerm(this)) { // non-rec skolem terms need to be specifically allowed
+      const Shell::SkolemTracker* st = synthMan->getSkolemTracker(functor());
+      if (st == nullptr) {
+         return false;
+      }
+      unsigned idx = 0;
+      if (!recAncestors->find(st->recFnId, idx) || (idx != st->constructorId)) {
+        return false;
+      }
+      return true;
+    }
   }
-  return true;
-}
 
-bool Term::computableOrVar() const {
-  if (!env.signature->getFunction(this->functor())->computable()) {
-    return false;
+  // Top functor is computable, now recurse.
+  unsigned* idx = nullptr;
+  if (synthMan->isRecTerm(this)) {
+    ALWAYS(recAncestors->getValuePtr(functor(), idx, -1));
   }
-  SubtermIterator sit(this);
-  while (sit.hasNext()) {
-    TermList t = sit.next();
-    if (t.isTerm() && !env.signature->getFunction(t.term()->functor())->computable()) {
-      return false;
+  for (unsigned i = 0; i < _arity; i++) {
+    const TermList* t = nthArgument(i);
```

Comment:
If you're not considering uncomputability of type arguments, I suggest looping up to `numTermArguments()` and using `termArg(i)`.